### PR TITLE
Feature/trade off table

### PR DESCRIPTION
**Pull request by chinook25**

### CODE REVIEW

File: app/ts/PreferencesTab/Preferences/TradeOff/TradeOff.tsx [hunk 1]
```diff
@@ -0,0 +1,47 @@
...
+              </Grid>
+              <Grid item xs={12}>
+                <TradeOffTable />
+              </Grid>
+            </Grid>
+          </TradeOffContextProviderComponent>
+        </Grid>
+      ) : (
+        <Grid item xs={12}>
+          All partial value functions need to be linear in order to explore
```
Comment (DanielReid) [inline app/ts/PreferencesTab/Preferences/TradeOff/TradeOff.tsx:41]: message not as specified in story

File: app/ts/PreferencesTab/Preferences/TradeOff/TradeOff.tsx [hunk 2]
```diff
@@ -0,0 +1,47 @@
...
+import TradeOffReferenceCriterion from './TradeOffReferenceCriterion/TradeOffReferenceCriterion';
+import TradeOffReferenceCriterionStatement from './TradeOffReferenceCriterionStatement/TradeOffReferenceCriterionStatement';
+import TradeOffTable from './TradeOffTable/TradeOffTable';
+import _ from 'lodash';
+import IPvf from '@shared/interface/Problem/IPvf';
+
+export default function TradeOff(): JSX.Element {
+  const {pvfs} = useContext(PreferencesContext);
+
+  const areAllPvfsLinear = _.every(pvfs, (pvf: IPvf): boolean => {
```
Comment (kzalite) [inline app/ts/PreferencesTab/Preferences/TradeOff/TradeOff.tsx:14]: closed version has hasNonLinearPvf function in PreferencesUtil. I think I like this version better but maybe would be nice to unify?

File: app/ts/PreferencesTab/Preferences/TradeOff/TradeOffContext/TradeOffContext.tsx [hunk 3]
```diff
@@ -0,0 +1,99 @@
...
+}): JSX.Element {
+  const {filteredCriteria} = useContext(SubproblemContext);
+  const {currentSubproblem} = useContext(WorkspaceContext);
+  const {currentScenario} = useContext(PreferencesContext);
+
+  const [referenceCriterion, setReferenceCriterion] = useState<ICriterion>(
+    filteredCriteria[0]
+  );
+  const [criteria, setCriteria] = useState<ICriterion[]>(
+    filteredCriteria.slice(1, filteredCriteria.length)
```
Comment (DanielReid) [inline app/ts/PreferencesTab/Preferences/TradeOff/TradeOffContext/TradeOffContext.tsx:26]: the second argument to `slice` is not needed

File: app/ts/PreferencesTab/Preferences/TradeOff/TradeOffContext/ITradeOffContext.ts [hunk 4]
```diff
@@ -0,0 +1,14 @@
+import ICriterion from '@shared/interface/ICriterion';
+
+export default interface ITradeOffContext {
+  criteria: ICriterion[];
+  lowerBound: number;
+  partOfInterval: number;
+  referenceCriterion: ICriterion;
+  upperBound: number;
+  value1: number;
```
Comment (kzalite) [inline app/ts/PreferencesTab/Preferences/TradeOff/TradeOffContext/ITradeOffContext.ts:9]: valueForm?

File: app/ts/PreferencesTab/Preferences/TradeOff/TradeOffContext/TradeOffContext.tsx [hunk 5]
```diff
@@ -0,0 +1,99 @@
...
+  );
+
+  const [
+    configuredLowerBound,
+    configuredUpperBound
+  ] = currentSubproblem.definition.ranges[referenceCriterion.dataSources[0].id];
+
+  const [lowerBound, setLowerBound] = useState<number>(configuredLowerBound);
+  const [upperBound, setUpperBound] = useState<number>(configuredUpperBound);
+  const [value1, setValue1] = useState<number>(
```
Comment (DanielReid) [inline app/ts/PreferencesTab/Preferences/TradeOff/TradeOffContext/TradeOffContext.tsx:36]: `initialReferenceValue`

File: app/ts/PreferencesTab/Preferences/TradeOff/TradeOffContext/ITradeOffContext.ts [hunk 6]
```diff
@@ -0,0 +1,14 @@
...
+  value2: number;
```
Comment (kzalite) [inline app/ts/PreferencesTab/Preferences/TradeOff/TradeOffContext/ITradeOffContext.ts:10]: valueTo?

File: app/ts/PreferencesTab/Preferences/TradeOff/TradeOffContext/TradeOffContext.tsx [hunk 7]
```diff
@@ -0,0 +1,99 @@
...
+    (configuredUpperBound - configuredLowerBound) * 0.45 + configuredLowerBound
+  );
+  const [value2, setValue2] = useState<number>(
```
Comment (DanielReid) [inline app/ts/PreferencesTab/Preferences/TradeOff/TradeOffContext/TradeOffContext.tsx:39]: `newReferenceValue`

File: app/ts/PreferencesTab/Preferences/TradeOff/TradeOffContext/TradeOffContext.tsx [hunk 8]
```diff
@@ -0,0 +1,99 @@
...
+    (configuredUpperBound - configuredLowerBound) * 0.55 + configuredLowerBound
+  );
+  const weight = currentScenario.state.weights.mean[referenceCriterion.id];
+  const [partOfInterval, setPartOfInterval] = useState<number>(
+    (value2 - value1) / ((configuredUpperBound - configuredLowerBound) * weight)
```
Comment (DanielReid) [inline app/ts/PreferencesTab/Preferences/TradeOff/TradeOffContext/TradeOffContext.tsx:44]: complex enough calc to split off to function, especially given reuse in line 52/53

File: app/ts/PreferencesTab/Preferences/TradeOff/TradeOffContext/TradeOffContext.tsx [hunk 9]
```diff
@@ -0,0 +1,99 @@
...
+
+  useEffect(() => {
+    const weight = currentScenario.state.weights.mean[referenceCriterion.id];
+    setPartOfInterval(
+      (value2 - value1) /
+        ((configuredUpperBound - configuredLowerBound) * weight)
+    );
+  }, [value1, value2]);
+
+  function initValues() {
```
Comment (DanielReid) [inline app/ts/PreferencesTab/Preferences/TradeOff/TradeOffContext/TradeOffContext.tsx:57]: I spent too much time thinking about why this function was redoing the initialisations that already happen in 36 etc.
Would have saved me if this function was called `reset()` :)

File: app/ts/PreferencesTab/Preferences/TradeOff/TradeOffContext/TradeOffContext.tsx [hunk 10]
```diff
@@ -0,0 +1,99 @@
...
+  children: any;
+}): JSX.Element {
+  const {filteredCriteria} = useContext(SubproblemContext);
+  const {currentSubproblem} = useContext(WorkspaceContext);
+  const {currentScenario} = useContext(PreferencesContext);
+
+  const [referenceCriterion, setReferenceCriterion] = useState<ICriterion>(
+    filteredCriteria[0]
+  );
+  const [criteria, setCriteria] = useState<ICriterion[]>(
```
Comment (DanielReid) [inline app/ts/PreferencesTab/Preferences/TradeOff/TradeOffContext/TradeOffContext.tsx:25]: `otherCriteria` makes the distinction from reference clearer

File: app/ts/PreferencesTab/Preferences/TradeOff/TradeOffReferenceCriterion/TradeOffReferenceCriterion.tsx [hunk 11]
```diff
@@ -0,0 +1,51 @@
...
+    event: ChangeEvent<{value: string}>
+  ): void {
+    updateReferenceCriterion(event.target.value);
+  }
+
+  function getReferenceCriterionOptions(): JSX.Element[] {
+    return _.map(
+      filteredCriteria,
+      (criterion: ICriterion): JSX.Element => {
+        return (
```
Comment (DanielReid) [inline app/ts/PreferencesTab/Preferences/TradeOff/TradeOffReferenceCriterion/TradeOffReferenceCriterion.tsx:24]: the function curlies and return statement are unneeded

File: app/ts/PreferencesTab/Preferences/TradeOff/TradeOffTable/TradeOffTable.tsx [hunk 12]
```diff
@@ -0,0 +1,89 @@
...
+  const {criteria, partOfInterval} = useContext(TradeOffContext);
+  const {pvfs, currentScenario} = useContext(PreferencesContext);
+
+  function getRows(): JSX.Element[] {
+    return _.map(
+      criteria,
+      (criterion: ICriterion): JSX.Element => {
+        const unit = criterion.dataSources[0].unitOfMeasurement;
+        const usePercentage = showPercentages && canBePercentage(unit.type);
+        const b = getB(criterion, usePercentage);
```
Comment (kzalite) [inline app/ts/PreferencesTab/Preferences/TradeOff/TradeOffTable/TradeOffTable.tsx:32]: improvedValue?

File: app/ts/PreferencesTab/Preferences/TradeOff/TradeOffTable/TradeOffTable.tsx [hunk 13]
```diff
@@ -0,0 +1,89 @@
...
+          </TableRow>
+        );
+      }
+    );
+  }
+
+  function getB(criterion: ICriterion, usePercentage: boolean): number {
+    const weight = currentScenario.state.weights.mean[criterion.id];
+    const pvf = pvfs[criterion.id];
+    const range = pvfs[criterion.id].range;
```
Comment (kzalite) [inline app/ts/PreferencesTab/Preferences/TradeOff/TradeOffTable/TradeOffTable.tsx:53]: or pvf.range? Not sure what I prefer :)

File: app/ts/PreferencesTab/Preferences/TradeOff/TradeOffReferenceCriterionStatement/TradeOffSlider/TradeOffSlider.tsx [hunk 14]
```diff
@@ -0,0 +1,82 @@
...
+
+  const unit = referenceCriterion.dataSources[0].unitOfMeasurement.type;
+  const usePercentage = showPercentages && canBePercentage(unit);
+
+  function openSlider(event: React.MouseEvent<HTMLButtonElement>): void {
+    setDialogOpen(true);
+    setAnchorElement(event.currentTarget);
+  }
+
+  function closeDialog(): void {
```
Comment (DanielReid) [inline app/ts/PreferencesTab/Preferences/TradeOff/TradeOffReferenceCriterionStatement/TradeOffSlider/TradeOffSlider.tsx:40]: `openSlider` `closeDialog`

File: app/ts/PreferencesTab/Preferences/TradeOff/TradeOffTable/TradeOffTable.tsx [hunk 15]
```diff
@@ -0,0 +1,89 @@
...
+                ? ''
+                : `This value is unrealistic given the criterion's range`}
+            </TableCell>
+          </TableRow>
+        );
+      }
+    );
+  }
+
+  function getB(criterion: ICriterion, usePercentage: boolean): number {
```
Comment (DanielReid) [inline app/ts/PreferencesTab/Preferences/TradeOff/TradeOffTable/TradeOffTable.tsx:50]: `calcCorrespondingValue(otherCriterion: ICriterion, usePercentage: boolean): number {`

File: app/ts/PreferencesTab/Preferences/TradeOff/TradeOffTable/TradeOffTable.tsx [hunk 16]
```diff
@@ -0,0 +1,89 @@
...
+    const weight = currentScenario.state.weights.mean[criterion.id];
+    const pvf = pvfs[criterion.id];
+    const range = pvfs[criterion.id].range;
+    const interval = range[1] - range[0];
+    const change = weight * partOfInterval * interval;
```
Comment (DanielReid) [inline app/ts/PreferencesTab/Preferences/TradeOff/TradeOffTable/TradeOffTable.tsx:55]: `change = (otherWeight / referenceWeight) * partOfInterval * interval`

assuming the weight is taken out of the interval calculation

**Review by kzalite — CHANGES_REQUESTED** [review]

File: app/ts/PreferencesTab/Preferences/TradeOff/TradeOffTable/TradeOffTable.tsx [hunk 17]
```diff
@@ -0,0 +1,89 @@
...
+    const modifier = usePercentage ? 100 : 1;
```
Comment (DanielReid) [inline app/ts/PreferencesTab/Preferences/TradeOff/TradeOffTable/TradeOffTable.tsx:56]: why not `getPercentifiedValue` like elsewhere?

File: app/ts/ScenarioSelection/ScenarioSelection.tsx [hunk 18]
```diff
@@ -19,13 +19,16 @@ export default function ScenarioSelection({
...
-    return _.map(scenarios, (scenario: IMcdaScenario) => {
-      return (
-        <option value={scenario.id} key={scenario.id}>
-          {scenario.title}
-        </option>
-      );
-    });
+    return _.map(
+      scenarios,
+      (scenario: IMcdaScenario): JSX.Element => {
```
Comment (DanielReid) [inline app/ts/ScenarioSelection/ScenarioSelection.tsx:24]: curlies and return not needed

**Review by DanielReid — CHANGES_REQUESTED** [review]

File: app/ts/PreferencesTab/Preferences/TradeOff/TradeOffContext/TradeOffContext.tsx (hunk 7 above)
Comment (chinook25) [reply app/ts/PreferencesTab/Preferences/TradeOff/TradeOffContext/TradeOffContext.tsx:39]: What do you mean?

File: app/ts/PreferencesTab/Preferences/TradeOff/TradeOffContext/TradeOffContext.tsx (hunk 5 above)
Comment (chinook25) [reply app/ts/PreferencesTab/Preferences/TradeOff/TradeOffContext/TradeOffContext.tsx:36]: What do you mean?

File: app/ts/PreferencesTab/Preferences/TradeOff/TradeOff.tsx (hunk 2 above)
Comment (chinook25) [reply app/ts/PreferencesTab/Preferences/TradeOff/TradeOff.tsx:14]: I also want to check here whether all pvfs are set.

File: app/ts/PreferencesTab/Preferences/TradeOff/TradeOffReferenceCriterionStatement/TradeOffReferenceCriterionStatement.tsx [hunk 19]
```diff
@@ -0,0 +1,55 @@
...
+    upperBound,
+    referenceValueFrom,
+    referenceValueTo,
+    setReferenceValueFrom,
+    setReferenceValueTo,
+    referenceCriterion
+  } = useContext(TradeOffContext);
+
+  const epsilon = 0.01 * (upperBound - lowerBound);
+  const stepSize1 = (referenceValueTo - epsilon - lowerBound) * 0.1;
```
Comment (DanielReid) [inline app/ts/PreferencesTab/Preferences/TradeOff/TradeOffReferenceCriterionStatement/TradeOffReferenceCriterionStatement.tsx:17]: still some x1, x2 hiding here

**Review by kzalite — APPROVED** [review]
nice job meme here

**Review by DanielReid — APPROVED** [review]

File: nightwatch.local.conf.js [hunk 20]
```diff
@@ -24,7 +24,7 @@ module.exports = {
       desiredCapabilities: {
         browserName: 'firefox',
         'moz:firefoxOptions': {
-          args: ['-headless']
+          // args: ['-headless']
```
Comment (kzalite) [inline nightwatch.local.conf.js:27]: usual leftover comment

File: test/endToEnd/tradeOffTest.js [hunk 21]
```diff
@@ -0,0 +1,116 @@
+"use strict";
```
Comment (kzalite) [inline test/endToEnd/tradeOffTest.js:1]: file should not be tracked

File: test/endToEnd/tradeOffTest.ts [hunk 22]
```diff
@@ -0,0 +1,124 @@
...
+
+function beforeEach(browser: NightwatchBrowser) {
+  browser.resizeWindow(1366, 728);
+  const workspaceTitle =
+    'Antidepressants - relative effectiveness analysis (Van Valkenhoef et al, J Clin Epi, 2012)';
+
+  loginService.login(browser);
+  workspaceService.cleanList(browser);
+  workspaceService.addExample(browser, workspaceTitle);
+  browser.expect.element('#workspace-0').text.to.equal(workspaceTitle);
```
Comment (kzalite) [inline test/endToEnd/tradeOffTest.ts:22]: :heart:

**Review by kzalite — CHANGES_REQUESTED** [review]

**Pull request merged**